src/main/java/com/google/sps/Scheduler.java

```diff
@@ -1,15 +1,113 @@
 package com.google.sps;
 
-import java.util.Collection;
-
 import com.google.sps.data.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 /**
- * Static class for permuting schedules given a list of courses.
+ * Static class for permuting schedules given a list of courses and a set of Invariants.
  */
 public final class Scheduler {
-    public Collection<Schedule> generateSchedules(Collection<Course> courses, Invariants invariants){
-        //TODO: This
-        return null;
+  /**
+   * Generates all valid schedules given a list of courses and invariant
+   * constraints.
+   * 
+   * @param courses    The list of courses to generate the schedules from.
+   * @param invariants The invariants that constrain the valid schedules
+   * @return A list of valid schedules.
+   */
+  public List<Schedule> generateSchedules(List<Course> courses, Invariants invariants) {
+    ArrayList<Course> betterCourses = new ArrayList<>(courses);
+    Collections.sort(betterCourses, Course.REQUIRED_TO_FRONT);
+    Set<List<Course>> schedulesSet = new HashSet<List<Course>>();
+
+    generateSchedulesHelper(betterCourses, new ArrayList<>(), invariants, schedulesSet);
+
+    System.out.println(courses);
+    List<Schedule> schedules = new ArrayList<>();
+    for (List<Course> schedule : schedulesSet) {
+      schedules.add(new Schedule(schedule));
+    }
+
+    return schedules;
+  }
+
+  /**
+   * Recursively iterates through a list of courses and adds schedules to the
+   * given set.
+   * 
+   * @param courses    The list of courses to iterate through
+   * @param schedule   The current schedule to be added/operated on
+   * @param invariants The invariants that constrain the valid schedules
+   * @param schedules  The set of schedules that is being generated/added to.
+   */
+  private void generateSchedulesHelper(List<Course> courses, List<Course> schedule, 
+      Invariants invariants, Set<List<Course>> schedules) {
+
+    // This means that we have a valid schedule, and should add it to the list.
+    float currCredits = totalCredits(schedule);
+    if (currCredits >= invariants.getMinCredits() && currCredits <= invariants.getMaxCredits()) {
```

Comment:
nit-adjacent: you could improve legibility here by implementing a `meetsCreditRequirements(float credits)` method on the `Invariants` class.

```
if (invariants.meetsCreditRequirements(credits)) {
  ...
}
```

Is pretty well self-documenting. It would also be easy to extend this method to be more of a "does this schedule meet all of the requirements" function and keep the logic checking for this well contained rather than have it grow into an increasingly dense `if` statement.